### ISSUE TITLE
Mesh refinement with Infinite elements fails

### ISSUE THREAD

**Issue by BalticPinguin** [libMesh/libmesh#1149]
Hi,
It is me again with some issue with infinite elements: I wanted to try to refine my mesh that has infinite elements but failed with the message 
'ERROR: Bad ElemType = INFPRISM6 for FIRST order approximation!'
which is thrown when trying to reinit the sides; more precisely, the backtrace is 
`
0: libMesh::print_trace(std::ostream&) 

1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: /home/tm162/bin/libmesh/my_libmesh/lib/libmesh_dbg.so.0(+0x12c9b50) [0x7ffff5ab7b50]
3: libMesh::FE<3u, (libMesh::FEFamily)0>::n_dofs(libMesh::ElemType, libMesh::Order)
4: libMesh::FE<3u, (libMesh::FEFamily)0>::n_shape_functions(libMesh::ElemType, libMesh::Order)
5: void libMesh::FEMap::init_reference_to_physical_map<3u>(std::__debug::vector<libMesh::Point, std::allocator<libMesh::Point> > const&, libMesh::Elem const*)
6: libMesh::FE<3u, (libMesh::FEFamily)0>::reinit(libMesh::Elem const*, std::__debug::vector<libMesh::Point, std::allocator<libMesh::Point> > const*, std::__debug::vector<double, std::allocator<double> > const*)
7: libMesh::JumpErrorEstimator::reinit_sides()
8: libMesh::JumpErrorEstimator::estimate_error(libMesh::System const&, libMesh::ErrorVector&, libMesh::NumericVector<std::complex<double> > const*, bool)
`
As I see it, JumpErrorEstimator::reinit_sides () calles the fe->reinit() directly which of course does not know the infinite element types.

My question now is mainly: Is mesh refinement supposed to work with infinite elements or do I need to expect there more errors with that? 
Best,
Hubert

**Comment by roystgnr** [comment]
On Mon, 31 Oct 2016, Hubert Weißmann wrote:

> My question now is mainly: Is mesh refinement supposed to work with infinite elements or do I need to expect there more errors with that?

That's two questions; the answer to both is "yes".

It's been years and years since we had a major developer doing work
with infinite elements, which means that the only thing between the
InfFE code and bit-rot has been our test coverage, of which we don't
have enough.  PRs adding unit tests would be greatly appreciated.

I'll take a crack at adding refinement to miscellaneous_ex1 and see
what I can trigger and fix.
---
Roy

**Comment by BalticPinguin** [comment]
Ok, that sounds great.
I didn't look closer at the refinement algorithm so far therefore I was wondering whether it would try to refine the 'infinite' region as well and therefore is not suitable at all.

**Comment by roystgnr** [comment]
Our h refinement should only split the "base" elements, while leaving the infinite region to be refined solely via increases in radial_order.

I managed to replicate your problem with ease, and fix some of the underlying problems with difficulty, but now I've hit a snag:

In System::projection_solution, I have no idea what to do with the solution DoFs corresponding to the interesting bits of the radial basis functions.  On an INFHEX8 with radial_order=3, for example, we have 1 DoF stored on each of the 4 vertices, which can just be copied or interpolated, and we have 3 DoFs stored on each of the 4 mid-edge nodes, which can be... what?

**Comment by roystgnr** [comment]
Gah.  I tried just falling back on the projection code for the non-obvious DoFs, but then that just hits one of the same failed assertions.

We need either a non-static version of FE::n_dofs or we need to add an optional radial_order argument.  Either way this is starting to dig very deeply into very old APIs; don't expect it to be working soon.

If you'd check out #1150 and make sure it's not a regression then I'd appreciate it; at least if we can merge that, the work will be half done whenever someone gets back to it in another week or another year.

**Comment by BalticPinguin** [comment]
I must admit I have not digged into these details so far; not the infinite elements nor the refinement algorithm.
So I have only little understandig of your changes but I can see if I can help you with #1150 tomorrow.

**Comment by BalticPinguin** [comment]
I did check your code finally. At least my application (that uses almost all of the functionality I looked at so far) runs without any complains or visible differences, and also from looking at the changes I don't see there any problem.

**Comment by jwpeterson** [comment]
Closing since it sounds like the problem was fixed.  Please feel free to reopen if necessary.

**Comment by roystgnr** [comment]
Not fixed, I'm afraid; I fixed some of the simple bugs that were blocking it from working, but there are still deep changes that would need to be made to get projections working on infinite elements.

@BalticPinguin - it occurs to me that you could probably use adaptivity *without* projections at this point.  That's less useful on a steady problem (where you wouldn't be able to use coarse solutions as a good initial guess for the fine solution) and it's grossly less useful on a transient problem (where you would have to put your adaptivity loop *outside* rather than inside your time stepping loop) but it might be better than nothing.

Anyway, I'm going to reopen this as a reminder; if I ever dig into the FE API issues that are preventing threaded Clough-Tocher elements from working, I can finish this fix at the same time.

**Comment by BalticPinguin** [comment]
Hi Roy,
I'd like to come back to this issue finally.

To start with an update: The error message reads still the same; so I guess there is not much change in the JumpErrorEstimator during those years (at least not in these functions).

I understand that you might have only little time for this issue; but maybe you can help me figuring out the necessary steps to go.
From your comments it looks as if you had(have?) already some code to fix at least parts of the problem. Probably you can share your ideas so I can try to implement it.

Best regards

**Comment by BalticPinguin** [comment]
Hi @roystgnr ,
I am currently trying to make progress on this issue again: My PR on the ``FEInterface::shape_ptr`` API was in this context and resolved the above-mentioned issue. But now I am stuck at the next point: 
```
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: /data/tm162/bin/DSK/master_libmesh-mpich/lib/libmesh_dbg.so.0(+0x2a8d762) [0x7f87a6d34762]
3: libMesh::FE<3u, (libMesh::FEFamily)0>::n_dofs(libMesh::ElemType, libMesh::Order)
4: libMesh::FE<3u, (libMesh::FEFamily)0>::n_shape_functions(libMesh::ElemType, libMesh::Order)
5: libMesh::FE<3u, (libMesh::FEFamily)0>::init_shape_functions(std::vector<libMesh::Point, std::allocator<libMesh::Point> > const&, libMesh::Elem const*)
6: libMesh::FE<3u, (libMesh::FEFamily)0>::reinit(libMesh::Elem const*, std::vector<libMesh::Point, std::allocator<libMesh::Point> > const*, std::vector<double, std::allocator<double> > const*)
7: libMesh::JumpErrorEstimator::reinit_sides()
8: libMesh::JumpErrorEstimator::estimate_error(libMesh::System const&, libMesh::ErrorVector&, libMesh::NumericVector<std::complex<double> > const*, bool)
```
So the reinit_sides() always calls the ``FE``-branch which does not find the ``InfFE``-type.
When I understand it correctly, the issue is that the objects are copied from the ``_side_fe_var``-objects of the ``FEMContext``-class which are initialized in the ``FEMContext::init_internal_data``-function via ``FEAbstract::build`` (instead of ``FEAbstract::build_InfFE`` which would be needed for InfFE).

To resolve this issue, it might be best to adjust the ``FEMContext``-class, I think?
My immediate idea is to reset the the internal ``FEAbstract``-pointers to new ``InfFE``-pointers when we set the element? 
But probably there are much more elegant ways or you foresee there some problems one might get with such a scheme; so I'd like to discuss the options before doing actual changes...

**Comment by roystgnr** [comment]
Hmm... I wish there were more elegant ways, but I think the point at which we diverged from elegance may have been when we made InfFE and FE separate objects  At *this* point, I think you're right that the place to tackle this is FEMContext.

But if we just have FEMContext build a load of InfFE objects in init_internal_data it's going to be a waste of time and cache space for 99% of users.  So I think you're right - we want to wait to build InfFE objects until we actually set _elem to an infinite element, then make sure we swap back (but leave the InfFEs saved for later) if we set _elem to a geometrically-finite element.

The catch there is that it won't work for any code that grabs an FEBase from the FEMContext *before* iterating over elements ... but if we have any code like that then it would be breaking on mixed dimensional meshes too, and we'd need to fix that regardless.
